catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/OperationsStorageSupport.java

```diff
@@ -0,0 +1,147 @@
+import ddf.catalog.source.IngestException;
+import ddf.catalog.source.SourceUnavailableException;
+import ddf.catalog.util.impl.Requests;
+
+public class OperationsStorageSupport {
+    private static final Logger LOGGER = LoggerFactory.getLogger(OperationsStorageSupport.class);
+
+    private static final Logger INGEST_LOGGER =
+            LoggerFactory.getLogger(Constants.INGEST_LOGGER_NAME);
+
+    //
+    // Injected propertiers
+    //
+    private final SourceOperations sourceOperations;
+
+    private final QueryOperations queryOperations;
+
+    private Historian historian;
+
+    public OperationsStorageSupport(SourceOperations sourceOperations,
+            QueryOperations queryOperations) {
+        this.sourceOperations = sourceOperations;
+        this.queryOperations = queryOperations;
+    }
+
+    public void setHistorian(Historian historian) {
+        this.historian = historian;
+    }
+
+    <T extends StorageRequest> T prepareStorageRequest(T storageRequest,
+            Supplier<List<ContentItem>> getContentItems)
+            throws IngestException, SourceUnavailableException {
+        validateStorageRequest(storageRequest, getContentItems);
+
+        queryOperations.setFlagsOnRequest(storageRequest);
+
+        if (Requests.isLocal(storageRequest) && (!sourceOperations.isSourceAvailable(
+                sourceOperations.getCatalog())
+                || !isStorageAvailable(sourceOperations.getStorage()))) {
+            String message = "Local provider is not available, cannot perform storage operation.";
+            SourceUnavailableException sourceUnavailableException = new SourceUnavailableException(
+                    message);
+            if (INGEST_LOGGER.isWarnEnabled()) {
+                INGEST_LOGGER.warn(message, sourceUnavailableException);
+            }
+            throw sourceUnavailableException;
+        }
+
+        return storageRequest;
+    }
+
+    void commitAndCleanup(StorageRequest storageRequest, Optional<String> historianTransactionKey,
+            HashMap<String, Path> tmpContentPaths) {
+        if (storageRequest != null) {
+            try {
+                sourceOperations.getStorage()
+                        .commit(storageRequest);
+                historianTransactionKey.ifPresent(historian::commit);
+            } catch (StorageException e) {
+                LOGGER.error("Unable to commit content changes for id: {}",
+                        storageRequest.getId(),
+                        e);
+                try {
+                    sourceOperations.getStorage()
+                            .rollback(storageRequest);
+                } catch (StorageException e1) {
+                    LOGGER.error("Unable to remove temporary content for id: {}",
+                            storageRequest.getId(),
+                            e1);
+                } finally {
+                    try {
+                        historianTransactionKey.ifPresent(historian::rollback);
```

Comment:
I guess I'm not clear how to extract cleanly to the Historian. I can ask the Historian to commit or rollback and have it check the key's value, but that would only be a one-line change in each of these two locations. I can't have the Historian manage the other storage calls and the exception processing, so I'm not sure how much improvement we'd get.